vessim/sil.py

```diff
@@ -25,76 +26,96 @@
+    def microgrid(self) -> Microgrid | None:
         return self._microgrid
 
-    def get_actor(self, actor: str) -> dict:
-        self._process_incoming_data()
-        assert self._actor_infos is not None
-        return self._actor_infos
-
-    def get_p_delta(self) -> float:
-        self._process_incoming_data()
-        assert self._p_delta is not None
+    @property
+    def p_delta(self) -> float:
         return self._p_delta
 
-    def set_event(self, category: str, value: Any) -> None:
-        if self._outgoing_events_queue is not None:
-            self._outgoing_events_queue.put(
-                {
-                    "category": category,
-                    "time": datetime.now(),
-                    "value": value,
-                }
-            )
+    @property
+    def e_delta(self) -> float:
+        return self._e_delta
 
-    def _add_microgrid_data(self, time: datetime, data: dict) -> None:
-        if self._incoming_data_queue is not None:
-            self._incoming_data_queue.put((time, data))
-
-    def _consume_events(self) -> Iterable[dict]:
-        if self._outgoing_events_queue is not None:
-            yield from _iterate_queue(self._outgoing_events_queue)
-
-    # TODO-now note that this should really be run periodically
-    def _process_incoming_data(self) -> None:
-        if self._incoming_data_queue is not None:
-            for time, data in _iterate_queue(self._incoming_data_queue):
-                self._microgrid = data.pop("microgrid", self._microgrid)
-                self._actor_infos = data.pop("actor_infos", self._actor_infos)
-                self._p_delta = data.pop("p_delta", self._p_delta)
+    def _get_ts_range(
+        self,
+        series: list[tuple[DatetimeLike, Any]],
+        start_time: Optional[DatetimeLike],
+        end_time: Optional[DatetimeLike],
+    ) -> list[tuple[DatetimeLike, Any]]:
+        start_idx = 0 if start_time is None else bisect_left(series, (start_time,))
+        end_idx = len(series) if end_time is None else bisect_right(series, (end_time,))
+        return series[start_idx:end_idx]
+
+    def get_microgrid_ts(
+        self, start_time: Optional[DatetimeLike] = None, end_time: Optional[DatetimeLike] = None
+    ) -> list[tuple[DatetimeLike, Microgrid]]:
+        with self._ts_lock:
+            ts = self._microgrid_ts.copy()
+        return self._get_ts_range(ts, start_time, end_time)
 
-    def _finalize(self) -> None:
-        assert self._outgoing_events_queue is not None
-        self._outgoing_events_queue.close()
-        self._outgoing_events_queue = None
-        assert self._incoming_data_queue is not None
-        self._incoming_data_queue.close()
-        self._incoming_data_queue = None
+    def get_actor(self, actor: str) -> dict:
+        return self._actor_infos[actor]
+
+    def get_actor_ts(
+        self, start_time: Optional[DatetimeLike] = None, end_time: Optional[DatetimeLike] = None
+    ) -> list[tuple[DatetimeLike, dict]]:
```

Comment:
The return signature could also be updated, to help with possible confusion :)
```suggestion
    ) -> list[tuple[DatetimeLike, dict[str, dict]]]:
```